portfolio/src/main/client/src/util/data.ts

```diff
@@ -0,0 +1,58 @@
+/** A function that responds to a signal changing. */
+type Handler<T> = (value: T, oldValue?: T) => void;
+
+/**
+ * A function that unregisters a @see {Handler}.
+ *
+ * @returns `true` if that handler was registered, `false` otherwise.
+ */
+type Cleanup = () => boolean;
+
+/**
+ * A data signal, an object which notifies subscribers when data is modified.
+ */
+export class DataSignal<T> {
+  private readonly _handlers: Set<Handler<T>> = new Set();
+  private _value: T;
+
+  public constructor(value: T) {
+    this._value = value;
+  }
+
+  public get value(): T {
+    return this._value;
+  }
+
+  public set value(value: T) {
+    const oldValue = this._value;
+    this._value = value;
+
+    for (const handler of this._handlers) {
+      handler(value, oldValue);
+    }
+  }
```

Comment:
Use of setters to have wide reaching side effects makes code hard to follow and bugs hard to locate. Setters and getters should have side effects that are local to the class and predictable. For a pattern like this, I would recommend using a verbose model (signal.updateValue(value) and signal.getValue()).